### DISCUSSION TITLE
How do I configure unflattening?

### DISCUSSION THREAD

**Discussion by cmanners** [riok/mapperly#1153] [Q&A]
I have an object that I need to unflatten. Can you please provide an example? Thank you. 🙂

**Comment by latonz** [comment]
See the [docs](https://mapperly.riok.app/docs/configuration/flattening/).
If it doesn't help you need to provide a little more details. How does your mapper definition and your classes look so far? What and how do you want to map?

**Reply by cmanners** [reply]
Thank you very much for the quick response.

My code base is quite complicated so, instead of posting some of that, I thought I should come up with a simpler example. Here is what I came up with:

- I have a simple source class:
```csharp
public class Source
{
    
    public string Id { get; set; }
    public string Name { get; set; }
    public string HomeStreetAddress { get; set; }
    public string HomeCity { get; set; }
    public string WorkStreetAddress { get; set; }
    public string WorkCity { get; set; }
}
```
- I am mapping it to two classes:
```csharp
public class User
{
    public string Id { get; set; }
    public string Name { get; set; }
    public Address HomeAddress { get; set; } = new Address();
    public Address WorkAddress { get; set; } = new Address();
}
```

```csharp
public class Address
{
    public string StreetAddress { get; set; }
    public string City { get; set; }
}
```
- My mapper is defined as follows:
```csharp
using Riok.Mapperly.Abstractions;

namespace Mapperly_Example;

[Mapper]
public partial class Mapper
{
    [MapProperty(new []{nameof(Source.HomeStreetAddress)}, new []{nameof(User.HomeAddress), nameof(Address.StreetAddress)})]
    [MapProperty(new []{nameof(Source.HomeCity)}, new []{nameof(User.HomeAddress), nameof(Address.City)})]
    [MapProperty(new []{nameof(Source.WorkStreetAddress)}, new []{nameof(User.WorkAddress), nameof(Address.StreetAddress)})]
    [MapProperty(new []{nameof(Source.WorkCity)}, new []{nameof(User.WorkAddress), nameof(Address.City)})]
    public partial User Map(Source source);
}
```

To my pleasant surprise, the mapper worked exactly as I expected. The generated mapper is the following:
```csharp
// <auto-generated />
#nullable enable
namespace Mapperly_Example
{
    public partial class Mapper
    {
        [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "3.4.0.0")]
        public partial global::Mapperly_Example.User Map(global::Mapperly_Example.Source source)
        {
            var target = new global::Mapperly_Example.User();
            target.Id = source.Id;
            target.Name = source.Name;
            target.HomeAddress.StreetAddress = source.HomeStreetAddress;
            target.HomeAddress.City = source.HomeCity;
            target.WorkAddress.StreetAddress = source.WorkStreetAddress;
            target.WorkAddress.City = source.WorkCity;
            return target;
        }
    }
}
```

So, long story short, the problem is in my code somewhere. My search continues.

Thank you again for the quick response.

**Reply by cmanners** [reply]
I added the following to my mapper:
```csharp
public static partial IQueryable<User> Map(IQueryable<Source> reports);
```
This modified the generated code to the following:
```csharp
// <auto-generated />
#nullable enable
namespace Mapperly_Example
{
    public partial class Mapper
    {
        [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "3.4.0.0")]
        public partial global::Mapperly_Example.User Map(global::Mapperly_Example.Source source)
        {
            var target = new global::Mapperly_Example.User();
            target.Id = source.Id;
            target.Name = source.Name;
            target.HomeAddress.StreetAddress = source.HomeStreetAddress;
            target.HomeAddress.City = source.HomeCity;
            target.WorkAddress.StreetAddress = source.WorkStreetAddress;
            target.WorkAddress.City = source.WorkCity;
            return target;
        }

        [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "3.4.0.0")]
        public static partial global::System.Linq.IQueryable<global::Mapperly_Example.User> Map(global::System.Linq.IQueryable<global::Mapperly_Example.Source> reports)
        {
#nullable disable
            return System.Linq.Queryable.Select(reports, x => new global::Mapperly_Example.User()
            {
                Id = x.Id,
                Name = x.Name,
            });
#nullable enable
        }
    }
}
```
In the generated mapper, the method did not map the address fields. How can I get it to map the address fields?

**Reply by latonz** [reply]
Aren't there a bunch of RMG* diagnostics reported?

**Reply by cmanners** [reply]
There are three errors:
1. RMG054 - Mapper class Mapperly_Example.Mapper contains 'static partial' method. Use only instance method or only static methods.
2. RMG016 - Cannot map to init only member path Mapperly_Example.Address.HomeAddress.StreetAddress
3. RMG016 - Cannot map to init only member path Mapperly_Example.Address.WorkAddress.StreetAddress

The errors only appeared after adding the IQueryable method (i.e., `public static partial IQueryable<User> Map(IQueryable<Source> reports);`)

I do not know how to resolve these errors.

**Reply by latonz** [reply]
* `RMG054`: Mapperly does not support mixed static and instance mapping methods in the same mapper. In your case the queryable mapping method seems to be static. The other method seems to be an instance method. Either set both to static or deckare both as instance methods.
* `RMG016`: Unflattening is unfortunately not yet supported in queryable projection mappings. This is reported by this diagnostic. Improvements are tracked in https://github.com/riok/mapperly/issues/643. Feel free to contribute.
